Jasmeen/add git rebase 812 (#868)
ACCESS-NRI/access-hive.org.au: README.md

The ACCESS-Hive Docs website hosts the documentation for ACCESS users: getting set up, running models and model evaluation.

# How to Contribute
### I. For New Users

If you wish to add documentation to the ACCESS-Hive Docs website check the [contribution guide](https://access-hive.org.au/about/contribute/) for instructions.

### II. For [ACCESS-NRI GitHub Organisation](https://github.com/ACCESS-NRI) members

The ACCESS-Hive Docs website is built using [Material for MkDocs](https://squidfunk.github.io/mkdocs-material/). For detailed information on markdown syntax and different features, please refer to [mkdocs-material documentation](https://squidfunk.github.io/mkdocs-material/reference/).

1. **CLONE**

    On [GitHub](https://github.com/ACCESS-NRI/access-hive.org.au), click on the **< > Code** button (top-right) and copy the SSH remote URL (starts with git@github.com). 
    
    Now, run the following command in your local terminal to clone the ACCESS-Hive GitHub repository:

    ```ruby
    $ git clone git@github.com:ACCESS-NRI/access-hive.org.au.git

>     ```

Author: Thanks @atteggiani ! I would like to leave ruby, as it does syntax coloring which is always helpful. I could not find a syntax colouring specific for the terminal though...

>     ```

Author: It's not a high importance task though, happy to leave syntax coloring as well unless needed.

>     ```

Reviewer: Yes I get it, but as you said, it's not a syntax highlighting for the shell commands, and sometimes it doesn't produce reliable highlighting. 
I don't think there is any syntax highlighting for shell commands (there is the `shell`/`sh` or `console` keyword but it does not work as expected), so I would simply remove it.

> 

Reviewer: Before creating a git branch there needs to be an issue related to the contribution, so it needs to be specified to raise an issue related to the contribution, if there isn't already one open.
Something like:
```suggestion
2. **Raise an issue about the contribution**
    Before working on a contribution, it is always good practice to have a GitHub issue detailing the content of the contribution and why it is important. 
    If not already present, open a [new issue](https://github.com/ACCESS-NRI/access-hive.org.au/issues/new?template=simple-issue-template.md) about your contribution.
```

2. **MAKE CHANGES LOCALLY**

    For making changes locally, open your preferred IDE (ex. VSCode). 
    And in the terminal, create a new branch to make your local changes and then push it to GitHub:

    ```ruby
    # Create a new branch, name it as - <name>/<issue-summary>-<issue-no>

    $ git checkout -b john/fix-title-800
    ```

    ```ruby
    # Push the branch to upstream

    $ git push --set-upstream origin john/fix-title-800  
    ```

> 3. **COMMIT AND PUSH CHANGES**
> 
>     Make your desired changes to the branch `john/fix-title-800`, and push the commits to remote: 
> 
>     ```ruby
>     # Stage the current local directory
> 
>     $ git add .
>     ```
> 
>     ```ruby
>     # Commit the changes, include linked issue ID and meaningful commit descriptions.
> 
>     # Avoid general descriptions, such as 'a fix', 'useful change', 'home page bug fixes' etc.
> 
>     $ git commit -m "(#800) Camel-cased hive title!"
>     ```
> 
>     ```ruby
>     # Push the changes to the remote of your branch
> 
>     $ git push origin john/fix-title-800 
> 
>     # or, if the remote is known, run 
>     
>     $ git push
>     ```

Reviewer: ```suggestion
3. **Make changes, commit and push to the remote repo**

    While you are in your new `git` branch, make your desired changes, commit your changes, and push the commits to the remote repo: 

    ```
    # Stage the desired files
    git add ...

    # Commit the changes, including a meaningful commit descriptions.
    # Avoid general descriptions, such as 'a fix', 'useful change', 'home page bug fixes' etc.
    git commit -m "Improved how-to-contribute instructions in the README."
    
    # Push the changes to the remote
    git push 
    ```
    
    > [!TIP]
    > The first that you push your new `git` branch to the remote, you will have to set its upstream with:
    > ```
    > git push --set-upstream origin jasmeen/improve-how-to-contribute-812
    > ```
    
    Repeat these steps as many times as you want, until you are satisfied with all the changes and you are ready for your contribution to be reviewed.
```

> 4. **BE IN SYNC WITH THE `development` BRANCH**

Reviewer: ```suggestion
4. **Keep your branch in sync with the `development` branch**
```

>     

Reviewer: ```suggestion
    Sometimes, while you make changes for your contribution, other commits can be added to the upstream (remote) `development` branch. 
    It is good practice to always keep your branch in sync with the latest updates of the `development` branch, before opening a Pull Request and have your contribution reviewed.
```

>     There are two main methods for integrating feature branch with the main branch:
> 
>     - Merge (creates a merge commit for every `git merge`)
>     - Rebase (fast-forward merge, and linear history) 
> 
>     Recently, we are transitioning to using `git rebase` as one of the steps in our development workflow.
>     This would give an opportunity to polish a feature branch before merging into the main, and also avoiding
>     additional merge commits in the project history.
> 

Reviewer: ```suggestion
```

I would omit this, it adds information not relevant for the purpose of syncing the branch in the preferred way (rebase). The only important part is the command that is detailed below.

>     For syncing our feature branch into the main branch, use the following 
>     commands: 
> 
>     ```ruby
>     # Skip this step if you are already working on feature branch 
>     git checkout john/fix-title-800
> 
>     # Moves the entire feature branch on the tip of main branch
>     git rebase origin main  # or git pull --rebase
>     ```

Reviewer: ```suggestion
    For syncing your feature branch with the upstream `development` branch, use the following 
    command: 

    ```
    git rebase origin/development jasmeen/improve-how-to-contribute-812
    ```
    
    > [!WARNING]
    > After rebasing, you might have to resolve conflicts.
    > Also, to push commits of a local branch after rebasing you might need to use `git push --force-with-lease` (`--force-with-lease` is a safer alternative to `--force`).
    > For these reasons, always use `git rebase` carefully. Don't hesitate to have a chat with the Hive Docs team if unsure :) 
```

> 5. **DEPLOYING WEBSITE PREVIEW**
> 
>     During development, preview the changes made using [MkDocs live preview server](https://access-hive.org.au/about/contribute/contribute_on_github/#deploying-website-preview).
> 

Reviewer: ```suggestion
5. **Deploy website preview**

    During development, you can preview the changes made using [MkDocs live preview server](https://access-hive.org.au/about/contribute/contribute_on_github/#deploying-website-preview).
```

I would also move this point before "Keep your branch in sync ..." as it's something that people might wanna do while they're applying changes.

Reviewer: What would you split in two different steps?

I think people might want to deploy the preview while they are updating their branch, to check for their updates (that's for example what I do). That is why I think it might be better placed as point number 4 (instead of 5).

Author: Thanks @atteggiani ! Might give it a discussion in the Hive docs meeting, I will leave this PR open till then.

Reviewer: Do you want to keep it open because it might be modified after your discussion?

> 6. **REVIEWING A FEATURE WITH THE PULL REQUEST**
>     
>     Once we are happy with the commits in the feature branch, [create a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request)
>     and GitHub will automatically build a preview of the documentation using
>     [GitHub Actions](https://docs.github.com/en/actions). 
> 
>     Feel free to assign `ACCESS-NRI/WebOps` team as reviewers.
> 
> > **Note** - The active branch is `development`. The `main` branch is only updated weekly 
> with automatic merging (from `development` branch only).
> 
> > **Caution**: Use `git rebase` carefully and never use it on the commits that have already 
> been merged in the main branch. It would appear that the project history is abruptly changed. 
> Don't hesitate to chat with the hive docs team if unsure... :) 
> 
> > _Reference: [Atlassian Rebase tutorial](https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase)_
> , _[Atlassian Merging vs Rebasing tutorial](https://www.atlassian.com/git/tutorials/merging-vs-rebasing)_
> 
> 
> ## Styling guidelines for developers
> 

Reviewer: ```suggestion
6. **Open a Pull Request**
    
    Once you are happy with your contribution, [create a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/proposing-changes-to-your-work-with-pull-requests/creating-a-pull-request) to have it reviewed by the ACCESS-Hive Docs team.
    Feel free to assign `ACCESS-NRI/WebOps` team as reviewers.

> [!Note]
> The ACCESS-Hive Docs repo's default branch is `development`. The `main` branch is automatically updated daily by merging the `development` branch.

## Styling guidelines for developers

```
